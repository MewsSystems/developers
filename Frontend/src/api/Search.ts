import axios, { AxiosPromise } from 'axios'
import { RequestConfig } from './requestConfig'
import { List } from 'model/api/List'
import { Movie } from 'model/api/Movie'

export class SearchApi {
  /**
   * Search
   *
   * @param query Pass a text query to search. This value should be URI encoded.
   * @param page Specify which page to query.
   * @param include_adult Choose whether to inlcude adult (pornography) content in the results.
   * @param region Specify a ISO 3166-1 code to filter release dates. Must be uppercase.
   * @param year
   * @param primary_release_year
   * @param config The AxiosRequestConfig properties not predefined by the generated API
   */

  static getSearchResults(
    query: string,
    page?: number,
    include_adult?: boolean,
    region?: string,
    year?: number,
    primary_release_year?: number,
    config?: RequestConfig
  ): AxiosPromise<List<Movie>> {
    return axios({
      ...config,
      method: 'GET',
      url: `/search/movies`,
      params: {
        query,
        page,
        include_adult,
        region,
        year,
        primary_release_year,
      },
    })
  }
}
