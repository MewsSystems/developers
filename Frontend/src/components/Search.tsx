import { Input } from 'components/Input'
import React, { useState, useCallback, useRef } from 'react'
import { useTranslation } from 'react-i18next'
import { useTrackEvent } from 'hooks/useTrackEvent'
import { EVENT_CATEGORY, EVENT_ACTION } from 'constants/tracking'

export interface SearchProps {
  value?: string
  onChange?: (value: string) => void
  onClear?: () => void
  onPressEnter?: (value: string) => void
  className?: string
}

export const Search: React.FC<SearchProps> = ({
  value = '',
  onChange,
  onClear,
  onPressEnter,
  className,
}) => {
  const { t } = useTranslation('search')
  const { setTrackEvent } = useTrackEvent()
  const [searchValue, setSearchValue] = useState(value)
  const inputRef = useRef<HTMLInputElement>()

  const handleChange = useCallback(
    ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
      setSearchValue(value)
      onChange && onChange(value)
    },
    [onChange]
  )

  const handleKeyDown = useCallback(
    ({ key }: React.KeyboardEvent<HTMLInputElement>) => {
      if (key === 'Enter' && onPressEnter) {
        onPressEnter(searchValue)

        setTrackEvent({
          eventCategory: EVENT_CATEGORY.SEARCH,
          eventAction: EVENT_ACTION.SEARCH.SUBMIT,
          eventLabel: searchValue,
        })
      }
    },
    [onPressEnter, searchValue, setTrackEvent]
  )

  const handleClear = useCallback(() => {
    setSearchValue('')

    setTrackEvent({
      eventCategory: EVENT_CATEGORY.SEARCH,
      eventAction: EVENT_ACTION.SEARCH.CLEAR,
    })

    onClear && onClear()

    inputRef.current && inputRef.current.focus()
  }, [onClear, setTrackEvent])

  return (
    <Input
      ref={inputRef}
      fullWidth
      allowClear
      inputSize="large"
      type="text"
      placeholder={t('placeholder')}
      value={searchValue}
      onClear={handleClear}
      onChange={handleChange}
      onKeyDown={handleKeyDown}
      className={className}
    />
  )
}
