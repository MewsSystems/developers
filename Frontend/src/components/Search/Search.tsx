import React, { useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import ReactPaginate from 'react-paginate';

import scrollToTop from '../../utils/scrollToTop';
import { RootReducer } from '../../reducers';
import { fetchTopRatedMovies, searchMovies, handleSearchChange } from '../../actions/search';
import config from '../../config';

import { Title, SearchResultsWrapper, SearchInput, MovieLink, MoviePosterImage } from './Styled';

const Search = () => {
  const dispatch = useDispatch();
  const loading = useSelector((state: RootReducer) => state.search.loading);
  const searchValue = useSelector((state: RootReducer) => state.search.value);
  const list = useSelector((state: RootReducer) => state.search.list);
  const pagination = useSelector((state: RootReducer) => state.search.pagination, shallowEqual);

  useEffect(() => {
    // there is a warning is browser console about this, but this is OK
    // I don't want to rerender on every loading change
    // it just controls first fetch and does not fetch when we come back from movie page
    if (loading) {
      dispatch(fetchTopRatedMovies());
    }
  }, [dispatch]);

  const onSearchChange = (event: React.ChangeEvent) => {
    event.preventDefault();
    const value = (event.target as HTMLInputElement).value;
    dispatch(handleSearchChange(value));
    if (value) {
      dispatch(searchMovies(value));
    } else {
      dispatch(fetchTopRatedMovies());
    }
  };

  const handlePageChange = ({ selected }) => {
    if (searchValue) {
      dispatch(searchMovies(searchValue, selected + 1));
    } else {
      dispatch(fetchTopRatedMovies(selected + 1));
    }
    scrollToTop();
  };

  return (
    <div>
      <SearchInput onChange={onSearchChange} placeholder="Type to search..." type="search" value={searchValue} />
      {!searchValue && <Title>Top Rated Movies</Title>}
      <SearchResultsWrapper>
        {list.map((it) => (
          <MovieLink key={it.id} to={`/movie/${it.id}`}>
            {it.poster_path ? (
              <MoviePosterImage src={`${config.IMAGE_BASE_URL}${it.poster_path}`} alt={it.title} title={it.title} />
            ) : (
              <div>{it.title}</div>
            )}
          </MovieLink>
        ))}
      </SearchResultsWrapper>

      <ReactPaginate
        containerClassName="Search-Page-Pagination"
        disableInitialCallback
        forcePage={pagination.page - 1}
        initialPage={0}
        marginPagesDisplayed={2}
        onPageChange={handlePageChange}
        pageCount={pagination.total}
        pageRangeDisplayed={5}
      />
    </div>
  );
};

export default Search;
