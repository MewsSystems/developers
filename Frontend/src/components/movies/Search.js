import React from 'react';

const Search = ({ change }) => {

  return (
    <div>
      <input
        id="search"
        type="search"
        placeholder="Search movies..."
        onChange={e => change(e.target.value)}/>
    </div>
  )
};

export default Search;



