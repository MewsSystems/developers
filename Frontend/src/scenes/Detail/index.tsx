import { SceneMeta } from 'components/SceneMeta'
import { RouteComponentProps } from '@reach/router'
import { useGetMovieDetail } from 'hooks/useGetMovieDetail'
import React from 'react'
import { HeroBanner } from 'components/HeroBanner'
import { Heading } from 'components/Heading'
import { COLORS } from 'constants/colors'
import styled from 'styled-components'
import { Credits } from './components/Credits'
import { Content } from 'components/Content'
import { ErrorContent } from 'components/ErrorContent'
import { useTranslation } from 'react-i18next'
import { Rating } from 'components/Rating'
import { Loading } from './components/Loading'
import { Info } from './components/Info'

const StyledHeroBanner = styled(HeroBanner)`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  padding: 2rem;
`

const Title = styled(Heading)`
  margin-bottom: 0.2rem;
  text-shadow: 1px 1px 0 ${COLORS.BLACK};
`

const Overview = styled.p`
  max-width: 45rem;
  margin: 1rem 0 0;
  color: ${COLORS.WHITE};
  text-shadow: 1px 1px 0 ${COLORS.BLACK};
`

const Grid = styled.div`
  display: grid;
  grid-gap: 2rem;
  grid-template-columns: auto 24rem;
`

export interface DetailProps extends RouteComponentProps {
  id?: number
}

export const Detail: React.FC<DetailProps> = ({ id }) => {
  const { t } = useTranslation()
  const { movieDetail, isLoading, hasErrored } = useGetMovieDetail(id)

  if (hasErrored)
    return (
      <ErrorContent
        title={t('error-content.title')}
        text={t('error-content.text')}
      />
    )
  if (isLoading) return <Loading />

  if (movieDetail) {
    const {
      backdrop_path,
      title,
      overview,
      vote_average,
      vote_count,
    } = movieDetail

    return (
      <>
        <SceneMeta title={title} description={overview} />
        <StyledHeroBanner background={backdrop_path && backdrop_path}>
          <Content>
            <Title level={1} color={COLORS.WHITE}>
              {title}
            </Title>
            <Rating rating={vote_average / 2} total={vote_count} />
            <Overview>{overview}</Overview>
          </Content>
        </StyledHeroBanner>
        <Content>
          <Grid>
            <Credits id={id} />
            <Info data={movieDetail} />
          </Grid>
        </Content>
      </>
    )
  }

  return null
}
