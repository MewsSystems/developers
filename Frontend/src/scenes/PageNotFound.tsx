import React from 'react'
import { RouteComponentProps } from '@reach/router'
import { SceneMetaProps, SceneMeta } from 'components/SceneMeta'
import { Heading } from 'components/Heading'
import { Content } from 'components/Content'

export const PageNotFound: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => (
  <>
    <SceneMeta title={title} description={description} />
    <Content>
      <Heading level={1}>Page not found</Heading>
    </Content>
  </>
)
