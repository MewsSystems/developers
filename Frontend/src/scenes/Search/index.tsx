import React, { useState, useCallback, useEffect } from 'react'
import { SceneMeta, SceneMetaProps } from 'components/SceneMeta'
import { State } from 'state/rootReducer'
import { useSelector } from 'react-redux'
import { RouteComponentProps, navigate } from '@reach/router'
import { Content } from 'components/Content'
import { Heading } from 'components/Heading'
import { useTranslation } from 'react-i18next'
import { COLORS } from 'constants/colors'
import styled from 'styled-components'
import { MovieList } from 'components/MovieList'
import { Movie } from 'model/api/Movie'
import { useGetSearchResults } from 'hooks/useGetSearchResults'
import { Spacing } from 'components/Spacing'
import { ErrorContent } from 'components/ErrorContent'
import { Loading } from './components/Loading'

const { PUBLIC_URL } = process.env

const SearchQuery = styled.span`
  color: ${COLORS.DARK_GRAY};
  font-weight: bold;
`
export const Search: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => {
  const { t } = useTranslation('search')
  const [movies, setMovies] = useState<Movie[]>([])
  const [page, setPage] = useState(1)
  const [totalPages, setTotalPages] = useState(0)
  const searchValue = useSelector((state: State) => state.search.value)
  const { results, isLoading, hasErrored } = useGetSearchResults(page)

  const handleLoadMore = useCallback(() => {
    setPage(page + 1)
  }, [page])

  const handleMovieClick = useCallback((id: number) => {
    navigate(`${PUBLIC_URL}/movie/${id}`)
  }, [])

  useEffect(() => {
    setMovies(state => [...state, ...results.results])
    setTotalPages(results.total_pages)
  }, [results])

  if (hasErrored)
    return (
      <ErrorContent
        title={t('error-content.title')}
        text={t('error-content.text')}
      />
    )

  if (isLoading) return <Loading />

  return (
    <>
      <SceneMeta title={`${title} ${searchValue}`} description={description} />
      <Content>
        <Heading level={1}>{t('search-results')}</Heading>
        {searchValue.length > 0 ? (
          <>
            <Spacing outer="0 0 2rem">
              {t('searching-for')} <SearchQuery>{searchValue}</SearchQuery>
            </Spacing>

            {results.total_results === 0 ? (
              <Heading level={2}>
                {t('no-search-results', { searchValue })}
              </Heading>
            ) : (
              <MovieList
                items={movies}
                onMovieClick={handleMovieClick}
                onLoadMore={handleLoadMore}
                page={page}
                totalPages={totalPages}
                loading={isLoading}
              />
            )}
          </>
        ) : (
          <Heading level={2}>{t('no-search-value')}</Heading>
        )}
      </Content>
    </>
  )
}
