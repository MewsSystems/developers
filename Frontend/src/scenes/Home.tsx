import React, { useEffect, useState, useCallback } from 'react'
import { RouteComponentProps, navigate } from '@reach/router'
import { SceneMetaProps, SceneMeta } from 'components/SceneMeta'
import { Heading } from 'components/Heading'
import { useTranslation } from 'react-i18next'
import { useThunkDispatch } from 'hooks/useThunkDispatch'
import { hydrateTrendingMovies } from 'state/actions/movies'
import { useSelector } from 'react-redux'
import { State } from 'state/rootReducer'
import { differenceInDays } from 'date-fns'
import { Movie } from 'model/api/Movie'
import { Content } from 'components/Content'
import { MovieList } from 'components/MovieList'

const { PUBLIC_URL } = process.env

export const Home: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => {
  const { t } = useTranslation('home')
  const dispatch = useThunkDispatch()
  const [trendingMovies, setTrendingMovies] = useState<Movie[]>([])
  const [page, setPage] = useState(1)
  const {
    updatedAt,
    trending: { results, total_pages },
    loading,
  } = useSelector((state: State) => state.movies)
  const dayDifference = differenceInDays(new Date(), new Date(updatedAt))

  const handleLoadMore = useCallback(() => {
    setPage(page + 1)
  }, [page])

  const reduceTrendingMovies = useCallback(() => {
    return Object.keys(results).reduce<Movie[]>((acc, index) => {
      if (Number(index) <= page) {
        acc = [...acc, ...results[index]]
      }

      return acc
    }, [])
  }, [results, page])

  const handleMovieClick = useCallback((id: number) => {
    navigate(`${PUBLIC_URL}/movie/${id}`)
  }, [])

  useEffect(() => {
    if (dayDifference > 7) {
      dispatch(hydrateTrendingMovies(true))
    } else {
      dispatch(hydrateTrendingMovies(false, page))
    }
  }, [dispatch, dayDifference, page])

  useEffect(() => {
    setTrendingMovies(reduceTrendingMovies())
  }, [page, results, reduceTrendingMovies])

  return (
    <>
      <SceneMeta title={title} description={description} />
      <Content>
        <Heading level={1}>{t('trending')}</Heading>
        <MovieList
          items={trendingMovies}
          page={page}
          totalPages={total_pages}
          loading={loading}
          onLoadMore={handleLoadMore}
          onMovieClick={handleMovieClick}
        />
      </Content>
    </>
  )
}
