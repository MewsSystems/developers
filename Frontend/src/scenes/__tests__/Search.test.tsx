import React from 'react'
import { Search } from 'scenes/Search'
import { shallowWithStore } from 'test/shallowWithStore'

describe('Test Search scene', () => {
  it('Tests snapshot', () => {
    expect(shallowWithStore(<Search />)).toMatchSnapshot()
  })
})
