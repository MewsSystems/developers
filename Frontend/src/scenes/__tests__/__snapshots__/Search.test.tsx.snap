// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Search scene Tests snapshot 1`] = `
<Fragment>
  <SceneMeta
    title="undefined "
  />
  <styled.div>
    <Heading
      level={1}
    >
      search-results
    </Heading>
    <Heading
      level={2}
    >
      no-search-value
    </Heading>
  </styled.div>
</Fragment>
`;
